File: examples/basic.c
/* Train a forest and score a batch: 100 inliers plus one outlier. */
#include <stdio.h>
#include <stdlib.h>
#include "iforest.h"

int main(void)
{
    int n = 101, m = 2, i;
    double *X = malloc(sizeof(double) * n * m);

    for (i = 0; i < 100; i++) {            /* 100 inliers */
        X[i * m + 0] = (double)rand() / RAND_MAX;
        X[i * m + 1] = (double)rand() / RAND_MAX;
    }
    X[100 * m + 0] = 5.0;                   /* 1 outlier */
    X[100 * m + 1] = 5.0;

    iforest_t f = iforest_train(X, n, m, 100, 0);   /* psi = 0 -> auto */
    double *s = malloc(sizeof(double) * n);
    iforest_score(f, X, n, m, s);
    printf("inlier %.3f   outlier %.3f\n", s[0], s[100]);

    iforest_free(f);
    free(X);
    free(s);
    return 0;
}

File: examples/basic.f90
! Train a forest and score a batch of points.
program basic
  use iforest, only: dp, IsolationForest, train_forest, predict_scores, free_forest
  implicit none
  type(IsolationForest) :: forest
  real(dp) :: X(100, 2), s(100)
  integer :: i
  real(dp) :: r1, r2

  do i = 1, 99
     call random_number(r1); call random_number(r2)
     X(i,:) = [r1, r2]                       ! inliers in the unit square
  end do
  X(100,:) = [5.0_dp, 5.0_dp]                ! one outlier

  call train_forest(forest, X, 100)          ! defaults: 100 trees, psi = min(256, n)
  call predict_scores(forest, X, 100, s)

  print '(a,f6.3)', "mean inlier score = ", sum(s(1:99)) / 99
  print '(a,f6.3)', "outlier score     = ", s(100)

  call free_forest(forest)
end program basic
